src/collector/Backend.cpp

```diff
@@ -121,16 +130,62 @@ void Backend::recalculate(uint64_t reserved_space)
         std::max(m_calculated.free_space - (m_calculated.total_space - m_calculated.effective_space), 0L);
 }
 
+void Backend::check_stalled(uint64_t stall_timeout_sec)
+{
+    time_t ts_now = time(nullptr);
+    if (ts_now <= time_t(m_stat.ts_sec)) {
+        m_calculated.stalled = false;
+        return;
+    }
+    m_calculated.stalled = ((ts_now - m_stat.ts_sec) > stall_timeout_sec);
+}
+
 void Backend::update_status()
 {
-    if (m_stat.error || m_stat.state != DNET_BACKEND_ENABLED || m_fs == nullptr)
+    if (m_stat.state != DNET_BACKEND_ENABLED) {
         m_calculated.status = STALLED;
-    else if (m_fs->get_status() == FS::BROKEN)
+        m_status_text = "Backend is not enabled";
+    } else if (m_calculated.stalled) {
+        m_calculated.status = STALLED;
+
+        std::ostringstream ostr;
+        ostr << "Backend statistics were gathered " << (time(nullptr) - m_stat.ts_sec) << " seconds ago";
+        m_status_text = ostr.str();
+    } else if (m_fs == nullptr) {
+        m_calculated.status = STALLED;
+
+        std::ostringstream ostr;
+        ostr << "Internal inconsistency: FS " << m_stat.fsid << " is not bound to backend " << m_node.get_key();
+        m_status_text = ostr.str();
+
+        BH_LOG(m_node.get_storage().get_app().get_logger(), DNET_LOG_ERROR, ostr.str().c_str());
+    } else if (m_fs->get_status() == FS::BROKEN) {
         m_calculated.status = BROKEN;
-    else if (m_read_only)
+        m_status_text = "Backend space limit is not properly configured on FS ";
+        m_status_text += m_fs->get_key();
+    } else if (m_stat.read_only) {
+        m_calculated.status = RO;
+        m_status_text = "Backend is switched to read-only state";
+    } else if (m_calculated.new_stat_commit_errors) {
         m_calculated.status = RO;
-    else
+        m_status_text = "Backend is read-only due to filesystem or disk errors";
+    } else {
         m_calculated.status = OK;
+        m_status_text = "Backend is OK";
+    }
+}
+
+bool Backend::group_changed() const
+{
+    if (m_group == nullptr)
+        return false;
+
+    return (m_group->get_id() != int(m_stat.group));
```

Comment:
RapidJSON sends event Uint64 for all numbers in our documents. I don't see the point of specifying word size to save for now.